packages/components/src/badge/index.tsx

```diff
@@ -0,0 +1,67 @@
+/**
+ * External dependencies
+ */
+import clsx from 'clsx';
+
+/**
+ * WordPress dependencies
+ */
+import { info, caution, error, published } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import type { BadgeProps } from './types';
+import Icon from '../icon';
+
+function Badge( {
+	className,
+	as: Component = 'div',
+	context = 'neutral',
+	children,
+	...props
+}: BadgeProps ) {
+	/**
+	 * Returns an icon based on the badge context.
+	 *
+	 * @return {JSX.Element | null} The corresponding icon for the provided context.
+	 */
+	function contextBasedIcon(): JSX.Element | null {
+		switch ( context ) {
+			case 'info':
+				return info;
+			case 'warning':
+				return caution;
+			case 'error':
+				return error;
+			case 'success':
+				return published;
+			default:
+				return null;
+		}
+	}
+
+	return (
+		<Component
+			className={ clsx(
+				'components-badge',
+				`components-badge--${ context }`,
+				context !== 'neutral' && 'components-badge--has-icon',
+				className
+			) }
+			aria-label={ `${ context }-badge` }
+			{ ...props }
+		>
+			{ context !== 'neutral' && (
+				<Icon
+					icon={ contextBasedIcon() }
+					size={ 20 }
```

Comment:
There aren't many (any?) examples of icons rendering at `20px` in the software, but I don't mind it here. What do you think @jasmussen?

<img width="160" alt="Screenshot 2024-12-02 at 13 28 30" src="https://github.com/user-attachments/assets/4b2fc732-b4df-4ff1-a5e7-5ea7a373e93b">

`24px` looks a bit odd next to the smaller text:

<img width="165" alt="Screenshot 2024-12-02 at 13 29 12" src="https://github.com/user-attachments/assets/2974e9fe-295b-4678-8d3e-f71f90ab8b8d">

Reply:
Hey, thanks for the ping, sorry for the belated response. 

I have semi strong opinions on this, all related to the icons being carefully designed to be displayed at 24x24 precisely, so that the _pixels render correctly_. Even if they are SVGs meant to scale, the pixels won't alias perfectly at any resolution. 

That's a long winded way of saying: if we are going to scale the icons down, we should do it according to some careful and deliberate math, so the pixels alias as well as possible. And we should do it very rarely: it should be an exception to the rule to use scaled icons, essentially, just like how we have heuristics for typography sizes. This badge can be such an exception, certainly.

So about that math: the icons are 24x24 with a 1.5px stroke width. At 20x20, the stroke width would be 1.25px. That would render blurry both at 1x screens, and at 2x screens (2.50px). If we scaled the icon to 16 (33% smaller icon), the stroke width would be exactly 1px, so it would be crisp at both 1x and 2x screens. 

How would 16x16 look in that context, can you try? If the math checks out, it should look very crisp, even if smaller. And smaller in this case would be good insofar as it looks intentional: it's smaller enough to be significant.

Reply:
Thanks for the context, we should document this, and maybe even restrict the `size` prop on `Icon` to specific values.

`16px` seems a bit small, but I'll take it over `24px`:

<img width="152" alt="Screenshot 2024-12-04 at 13 51 39" src="https://github.com/user-attachments/assets/280fbf50-cd48-428d-9d5e-244064039d0f">

Reply:
I quite like that, because it makes hte icon feel like it's part of the text, almost like an in-text glyph rather than an icon. But curious what others think.

Reply:
> Would it make sense to add a set of icons to the library specifically for badges at this smaller size? These could have adjusted stroke widths and proportions to better suit their context.

This is a bit of a large undertaking, and even now we are missing a build process (#38850) so this is likely not feasible in the foreseeable future. In that vein, I'd either keep the icons at 16px size, try 18px, just use 24x24, or omit the icon entirely.

Reply:
I vote we use the current icon at 16px. The resulting strokes are exactly 1px which I agree is a little thin, but should render sharply. 

In the future it might be beneficial to redraw the icons to use stroke-width, then we can be a bit more precise/flexible. I discuss this a bit in https://github.com/WordPress/gutenberg/issues/65786.

Reply:
> The resulting strokes are exactly 1px which I agree is a little thin, but should render sharply.

Agreed. This is especially important for 1x resolution monitors, which is still [the overwhelming majority of desktop monitors](https://gs.statcounter.com/screen-resolution-stats/).

Reply:
> In the future it might be beneficial to redraw the icons to use stroke-width, then we can be a bit more precise/flexible. I discuss this a bit in https://github.com/WordPress/gutenberg/issues/65786.

This would also potentially open the door for some animation fun. Agreed this is a big undertaking.